File: python/implementations/triangle.py
"""A naive approach to kNN, running in O(mndk) time."""

import math

from base import Base


class TriangleInequality(Base):

    def __init__(self, X_train, y_train, distance=None):
        super().__init__(X_train, y_train, distance)
        # Pre-computation of training distance matrix
        self.train_dist = [
            [
                math.sqrt(self.distance(self.X_train[i], self.X_train[j]))
                for j in range(i + 1, self.N)
            ]
            for i in range(self.N - 1)
        ]

    def predict(self, X_test, K):
        """See Base.predict"""
        pred = []
        for x0 in X_test:
            possible = [i for i in range(self.N)]
            curr_neighbours = [None for i in range(K)]

            i = 0
            n = self.N
            while i < n:
                p = possible[i]
                x = self.X_train[p]
                y = self.y_train[p]
                d = math.sqrt(self.distance(x0, x))
                if i >= K:
                    j = i + 1
                    while j < n:
                        if abs(d - self.train_dist[p][possible[j] - (p + 1)]) >\
                                curr_neighbours[K - 1][3]:
                            del possible[j]
                            n -= 1
                        else:
                            j += 1
                for k in range(K):
                    if curr_neighbours[k] is None or \
                            curr_neighbours[k][3] > d:
                        curr_neighbours.insert(k, (i, x, y, d))
                        del curr_neighbours[K]
                        break
                i += 1

            pred.append(sum(e[2] for e in curr_neighbours) / K)
        return pred

File: java/Triangle.java
// A naive approach to kNN, running in O(mndk) time.

import java.lang.Math;
import java.util.Arrays;
import java.util.Comparator;

public class Triangle extends Base {

  private double[][] train_dist;

  public Triangle(double[][] X_train, int[] y_train) {
    super(X_train, y_train);

    for(int i=0; i<(X_train.length-1); i++) {
      for(int j=0; j<i+1; j++) {
        this.train_dist[i][j] = this.dist(X_train[i], X_train[j]);
      }
    }
  }

  public double[] predict(double[][] X_test, int K) {

    double[] pred = new double[0];

    for(int j=0; j<X_test.length; j++) {
      boolean[] possible = new boolean[this.N];
      for(int e=0; e<possible.length; e++) {possible[e] = true;}
      int[] curr_i = new int[K];
      double[][] curr_x = new double[K][this.P];
      double[] curr_y = new double[K];
      double[] curr_d = new double[K];

      for(int i=0; i<this.X_train.length; i++) {
        if (!possible[i]) {continue;}
        double d = this.dist(X_test[j], this.X_train[i]);

        if(i>=K) {
          for(int l=0; l<(N-(i+1)); l++) {
            if (!possible[l]) {continue;}
            if (Math.abs(d - this.train_dist[i][l]) > curr_d[K-1]) {
              possible[i + l + 1] = false;
            }
          }
        }

        for(int k=0; k<K; k++) {
          int[] test_i = new int[K];
          double[][] test_x = new double[K][this.P];
          double[] test_y = new double[K];
          double[] test_d = new double[K];
          if(((curr_i[k] == test_i[k]) &&
              Arrays.equals(curr_x[k], test_x[k]) &&
              (curr_y[k] == test_y[k]) &&
              (curr_d[k] == test_d[k])) ||
             curr_d[k] > d) {
               curr_i[k] = i;
               curr_x[k] = this.X_train[i];
               curr_y[k] = this.y_train[i];
               curr_d[k] = d;

               // this is the irritating way to delete elements of an array in Java :(

               int[] temp_i = new int[K-1];
               double[][] temp_x = new double[K-1][this.P];
               double[] temp_y = new double[K-1];
               double[] temp_d = new double[K-1];
               for(int t=0; t<curr_i.length-1; t++) {
                 temp_i[t] = curr_i[t];
                 temp_x[t] = curr_x[t];
                 temp_y[t] = curr_y[t];
                 temp_d[t] = curr_d[t];
               }
               curr_i = temp_i;
               curr_x = temp_x;
               curr_y = temp_y;
               curr_d = temp_d;
               break;
             }
           }
        }

        double sum = 0;
        for (int e=0; e<curr_y.length; e++) {
          sum += curr_y[e];
        }
        sum /= K;

        // more appending incoming

        double[] temp = new double[(pred.length + 1)];
        for(int t=0; t<pred.length; t++) {
          temp[t] = pred[t];
        }
        temp[pred.length] = sum;
        pred = temp;
      }
      return pred;
  }
}
